server/root.go

```diff
@@ -24,35 +32,180 @@ import (
+	return func(w http.ResponseWriter, r *http.Request) {
+		// Wrap the ResponseWriter to remember the status
+		rww := &rpcserver.ResponseWriterWrapper{
+			Status:         -1,
+			ResponseWriter: w,
+		}
+		begin := time.Now()
+
+		rww.Header().Set("X-Server-Time", fmt.Sprintf("%v", begin.Unix()))
+
+		defer func() {
+			// Send a 500 error if a panic happens during a handler.
+			// Without this, Chrome & Firefox were retrying aborted ajax requests,
+			// at least to my localhost.
+			if e := recover(); e != nil {
+
+				// If RPCResponse
+				if res, ok := e.(rpctypes.RPCResponse); ok {
+					rpcserver.WriteRPCResponseHTTP(rww, res)
+				} else {
+					// For the rest,
+					logger.Error(
+						"Panic in RPC HTTP handler", "err", e, "stack",
+						string(debug.Stack()),
+					)
+					rpcserver.WriteRPCResponseHTTPError(rww, http.StatusInternalServerError, rpctypes.RPCInternalError(rpctypes.JSONRPCStringID(""), e.(error)))
+				}
+			}
+
+			// Finally, log.
+			durationMS := time.Since(begin).Nanoseconds() / 1000000
+			if rww.Status == -1 {
+				rww.Status = 200
+			}
+			logger.Info("Served RPC HTTP response",
+				"method", r.Method, "url", r.URL,
+				"status", rww.Status, "duration", durationMS,
+				"remoteAddr", r.RemoteAddr,
+			)
+		}()
+
+		handler.ServeHTTP(rww, r)
+	}
+}
+
+func startRPCServer(shutdownCtx ctx.Context, listener net.Listener, handler http.Handler, logger tmLog.Logger, cfg *rpcserver.Config) error {
+	logger.Info(fmt.Sprintf("Starting RPC HTTP server on %s", listener.Addr()))
+	recoverHandler := recoverAndLog(maxBytesHandler{h: handler, n: cfg.MaxBodyBytes}, logger)
+
+	s := &http.Server{
+		Handler: http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			ctx := r.Context()
+
+			select {
+			case <-ctx.Done():
+				fmt.Println("graceful handler exit")
+				rw.WriteHeader(http.StatusOK)
+				return
+			default:
+				recoverHandler(rw, r)
+			}
+
+		}),
+		ReadTimeout:    cfg.ReadTimeout,
+		WriteTimeout:   cfg.WriteTimeout,
+		MaxHeaderBytes: cfg.MaxHeaderBytes,
+		BaseContext: func(_ net.Listener) ctx.Context {
+			return shutdownCtx
+		},
+	}
+
+	g := new(errgroup.Group)
+	g.Go(func() error {
+		return s.Serve(listener)
+	})
+
+	g.Go(func() error {
+		// wait for interrupt signal coming from mainCtx
+		// and then go to server shutdown
+		<-shutdownCtx.Done()
+		ctx, cancel := ctx.WithTimeout(ctx.Background(), shutdownTimeout)
+		defer cancel()
+
+		return s.Shutdown(ctx)
+	})
+
+	if err := g.Wait(); err != nil {
+		logger.Info("RPC HTTP server stopped", "err", err)
+		return err
 	}
 
-	return err
+	return nil
 }
 
 // ServeCommands will generate a long-running rest server
 // (aka Light Client Daemon) that exposes functionality similar
 // to the cli, but over rest
-func ServeCommands(cdc *codec.Codec, registerRoutesFn func(*lcd.RestServer)) *cobra.Command {
+func ServeCommands(shutdownCtx ctx.Context, cdc *codec.Codec, registerRoutesFn func(ctx client.CLIContext, mux *mux.Router)) *cobra.Command {
```

Comment:
I do not think we should pass this ```shutdownCtx``` to each command because the shutdown signal should be something specific to each execution of a command, this is, something to be created inside the ```Run``` command functions. 

Could we do:
``` 
RunE: func(cmd *cobra.Command, args []string) error {
	shutdownCtx, stop := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
	defer stop()
 	err := StartRestServer(shutdownCtx, cdc, registerRoutesFn, restCh)
}
```

Reply:
In this case, each command would have created a context in its scope, where needed of course?

I could be wrong, but when we run our binary, each cmd should execute with the same context, since our main function is responsible for each execution of our defined commands?

Threading context down is a standard mechanism, not sure what I'm getting wrong here.